packages/jsxstyle/jsxstyle.tsx

```diff
@@ -1,21 +1,41 @@
 import * as React from 'react';
 import { componentStyles, getStyleCache } from 'jsxstyle-utils';
+import { StyleCache } from 'jsxstyle-utils/src/getStyleCache';
 
-export const cache = getStyleCache();
-
-function factory(displayName, defaultProps, tagName) {
-  tagName = tagName || 'div';
-  if (
-    process.env.NODE_ENV === 'development' &&
-    (typeof displayName !== 'string' || !displayName)
-  ) {
-    throw new Error(
-      'makeReactStyleComponentClass expects param 1 to be a valid displayName'
-    );
-  }
-
-  return class extends React.Component {
-    constructor(props) {
+import { CSSProperties } from './cssproperties';
+
+export const cache: StyleCache = getStyleCache();
+
+export interface ComponentPropProps {
+  className?: any;
+  style?: CSSProperties;
+  [key: string]: any;
+}
+
+export type ComponentProp =
+  | keyof JSX.IntrinsicElements
+  | React.ComponentClass<ComponentPropProps>
+  | React.SFC<ComponentPropProps>;
+
+export interface JsxstyleProps extends CSSProperties {
+  className?: string;
+  component?: ComponentProp;
+  mediaQueries?: { [key: string]: string };
+  props?: { [key: string]: any };
```

Comment:
`props` on the public interface really shouldn't be a `{[key: string]: any}`. How about something like this:
```ts
export type AnyComponent<ComponentProps> =
  | keyof JSX.IntrinsicElements
  | React.ComponentClass<ComponentProps>
  | React.SFC<ComponentProps>;

export type JsxstyleProps<ComponentProps> = {
  className?: string;
  component?: AnyComponent<ComponentProps>;
  mediaQueries?: {[key: string]: string};
  props?: ComponentProps;
  style?: CSSProperties;
} & CSSProperties;
```

Reply:
@steadicat I can dig it. The only thing I’m not sure about is what I pass to `JsxstyleProps` when I specify it as a type. For example:
```typescript
constructor(props: JsxstyleProps<???>) { // <--- what goes there?
  super(props);
  this.component = props.component || tagName;
  this.className = cache.getClassName(props, props.className);
}```

Sorry, this is probably basic stuff. Still learning this TypeScript biz.

Reply:
The generic param to `JsxstyleProps` should bubble all the way up to the caller. So, e.g., you’d define `Block` like this:

```ts
class Block<InnerComponentProps> extends React.Component<JsxstyleProps<InnerComponentProps>> {
  constructor(props: JsxstyleProps<InnerComponentProps>) {
    ...
  }
  ...
}
```

This way each call site determines which props are allowed. Ideally `InnerComponentProps` will be inferred by the value of the `component` prop. Should happen magically with the type definition I gave you above. Makes sense?

Reply:
Ahhhh yep that makes sense. This is cool!

Reply:
@steadicat does it make sense for `AnyComponent`’s props to extend `ComponentPropProps`? I’d like TypeScript to complain if a component specified as the `component` prop doesn’t accept the `className` prop (and perhaps the `style` prop as well).

Example:

```typescript
export interface ComponentPropProps {
  className?: any;
  style?: CSSProperties;
  [key: string]: any;
}

export type AnyComponent<Props extends ComponentPropProps> =
  | keyof JSX.IntrinsicElements
  | React.ComponentClass<Props>
  | React.SFC<Props>;

// ...
```

I think this does what I expect, but I might be wrong. React might already set `style` to `React.CSSProperties` and `className` to something like `any`? I’m not 100% sure.

Reply:
Yes! You got this! The index accessor on `ComponentPropProps` shouldn't be needed. Just specify the props you need the component to accept. I believe `className` is something like `string | null | undefined | false`. VSCode is great when you're discovering this stuff. You can hover over anything to see the inferred type, and jump to definition to see the full type declaration. Also, nit: the type of `style?: CSSProperties` should probably be the native React one, not the Jsxstyle kind.